File: src/orchestrator-cycles.ts
import type { CuriousState, CycleRecord } from "./types.js";

/** Full develop→review→sync rounds completed since the run started. */
export function completedCyclesSince(
  stateCycle: number,
  cycleAtStart: number,
): number {
  return stateCycle - cycleAtStart;
}

/** Stop after `--cycles N` when N full rounds finished and phase wrapped to develop. */
export function shouldStopAfterRequestedCycles(
  state: Pick<CuriousState, "phase" | "cycle">,
  cycleAtStart: number,
  cyclesLimit: number | undefined,
): boolean {
  if (cyclesLimit === undefined) {
    return false;
  }
  return (
    state.phase === "develop" &&
    completedCyclesSince(state.cycle, cycleAtStart) >= cyclesLimit
  );
}

/** In `--cycles` mode, abort the outer loop when a phase left lastError set. */
export function shouldAbortCyclesModeOnPhaseError(
  cyclesLimit: number | undefined,
  lastError: string | undefined,
): boolean {
  return cyclesLimit !== undefined && lastError !== undefined;
}

/** Stop when config maxCycles is positive and the state cycle reached the cap. */
export function shouldStopAtConfigMaxCycles(
  stateCycle: number,
  configMaxCycles: number,
): boolean {
  return configMaxCycles > 0 && stateCycle >= configMaxCycles;
}

/** Failed or cancelled runs keep the current phase for retry on the next iteration. */
export function retainsPhaseAfterRun(status: CycleRecord["status"]): boolean {
  return status !== "finished";
}

File: python/src/curious/orchestrator_cycles.py
from __future__ import annotations

from curious.types import CuriousState


def completed_cycles_since(state_cycle: int, cycle_at_start: int) -> int:
    return state_cycle - cycle_at_start


def should_stop_after_requested_cycles(
    state: CuriousState, cycle_at_start: int, cycles_limit: int | None
) -> bool:
    if cycles_limit is None:
        return False
    return (
        state.phase == "develop"
        and completed_cycles_since(state.cycle, cycle_at_start) >= cycles_limit
    )


def should_abort_cycles_mode_on_phase_error(
    cycles_limit: int | None, last_error: str | None
) -> bool:
    return cycles_limit is not None and last_error is not None


def should_stop_at_config_max_cycles(state_cycle: int, config_max_cycles: int) -> bool:
    return config_max_cycles > 0 and state_cycle >= config_max_cycles
